src/ring_theory/hahn_series.lean

```diff
@@ -0,0 +1,399 @@
+
+instance : has_zero (hahn_series Γ R) :=
+⟨{ coeff := 0,
+   is_wf_support' := by simp }⟩
+
+instance : inhabited (hahn_series Γ R) := ⟨0⟩
+
+@[simp]
+lemma zero_coeff {a : Γ} : (0 : hahn_series Γ R).coeff a = 0 := rfl
+
+@[simp]
+lemma support_zero : support (0 : hahn_series Γ R) = ∅ := function.support_zero
+
+/-- `single a r` is the Hahn series which has coefficient `r` at `a` and zero otherwise. -/
+def single (a : Γ) (r : R) : hahn_series Γ R := {
+  coeff := pi.single a r,
+  is_wf_support' := (set.is_wf_singleton a).mono pi.support_single_subset }
+
+variables {a b : Γ} {r : R}
+
+@[simp]
+theorem single_coeff_same : (single a r).coeff a = r := pi.single_eq_same a r
+
+@[simp]
+theorem single_coeff_of_ne (h : b ≠ a) : (single a r).coeff b = 0 := pi.single_eq_of_ne h r
+
+theorem single_coeff : (single a r).coeff b = if (b = a) then r else 0 :=
+by { split_ifs with h; simp [h] }
+
+@[simp]
+lemma support_single_of_ne (h : r ≠ 0) : support (single a r) = {a} :=
+pi.support_single_of_ne h
+
+@[simp]
+lemma single_eq_zero : (single a (0 : R)) = 0 := ext _ _ (pi.single_zero _)
+
+end zero
+section addition
+
+variable [linear_order Γ]
+
+instance [add_monoid R] : has_add (hahn_series Γ R) :=
+{ add := λ x y, { coeff := x.coeff + y.coeff,
+                  is_wf_support' := (x.is_wf_support.union y.is_wf_support).mono
+                    (function.support_add _ _) } }
+
+instance [add_monoid R] : add_monoid (hahn_series Γ R) :=
+{ zero := 0,
+  add := (+),
+  add_assoc := λ x y z, by { ext, apply add_assoc },
+  zero_add := λ x, by { ext, apply zero_add },
+  add_zero := λ x, by { ext, apply add_zero } }
+
+@[simp]
+lemma add_coeff' [add_monoid R] {x y : hahn_series Γ R} :
+  (x + y).coeff = x.coeff + y.coeff := rfl
+
+lemma add_coeff [add_monoid R] {x y : hahn_series Γ R} {a : Γ} :
+  (x + y).coeff a = x.coeff a + y.coeff a := rfl
+
+instance [add_comm_monoid R] : add_comm_monoid (hahn_series Γ R) :=
+{ add_comm := λ x y, by { ext, apply add_comm }
+  .. hahn_series.add_monoid }
+
+instance [add_group R] : add_group (hahn_series Γ R) :=
+{ neg := λ x, { coeff := λ a, - x.coeff a,
+                is_wf_support' := by { rw function.support_neg,
+                  exact x.is_wf_support }, },
+  add_left_neg := λ x, by { ext, apply add_left_neg },
+  .. hahn_series.add_monoid }
+
+@[simp]
+lemma neg_coeff' [add_group R] {x : hahn_series Γ R} : (- x).coeff = - x.coeff := rfl
+
+lemma neg_coeff [add_group R] {x : hahn_series Γ R} {a : Γ} : (- x).coeff a = - x.coeff a := rfl
+
+@[simp]
+lemma sub_coeff' [add_group R] {x y : hahn_series Γ R} :
+  (x - y).coeff = x.coeff - y.coeff := by { ext, simp [sub_eq_add_neg] }
+
+lemma sub_coeff [add_group R] {x y : hahn_series Γ R} {a : Γ} :
+  (x - y).coeff a = x.coeff a - y.coeff a := by simp
+
+instance [add_comm_group R] : add_comm_group (hahn_series Γ R) :=
+{ .. hahn_series.add_comm_monoid,
+  .. hahn_series.add_group }
+
+end addition
+section smul
```

Comment:
I don't understand if the section smul should really begin here and not after the end of distrib_mul_action